parlai/agents/transformer/polyencoder.py

```diff
@@ -137,6 +163,50 @@ def _set_text_vec(self, *args, **kwargs):
             obs['added_start_end_tokens'] = True
         return obs
 
+    def batchify(self, obs_batch: List[Message], sort: bool = False) -> Batch:
+        """
+        Override to handle image features.
+        """
+        batch = super().batchify(obs_batch, sort)
+
+        def _process_features(features: torch.Tensor) -> torch.Tensor:
+            assert features.size() == (self.image_features_dim,)
+            if self.use_cuda:
+                features = features.cuda()
+            if self.opt.get('fp16'):
+                features = features.half()
+            else:
+                features = features.float()
+
+            return features
+
+        if self.use_image_features:
+
+            # Checks/formatting of batch.image
+            bsz = batch.text_vec.size(0)
+            if batch.image is None or len(batch.image) == 0:
+                batch.image = [None] * bsz
+            else:
+                assert len(batch.image) == bsz
```

Comment:
hmm so this assumes that we only train with batches of _all_ images or _no_ images right? are we ruling out mixed batches?

Reply:
We can handle mixed batches: that's what the block with the `# Process all image feature vectors, or add in zero vectors if missing` is doing, filling in image features for specific examples that are missing them